File: python-sdk/hsafa/sse.py
import json
import threading
import time
from typing import Any, Callable, Dict, List, Optional

import requests


class SSEStream:
    """
    Server-Sent Events stream that runs in a background daemon thread.
    Auto-reconnects with exponential backoff on failure.
    """

    def __init__(
        self,
        url: str,
        headers: Dict[str, str],
        reconnect: bool = True,
        reconnect_delay: float = 1.0,
        max_reconnect_delay: float = 30.0,
        on_open: Optional[Callable[[], None]] = None,
        on_error: Optional[Callable[[Exception], None]] = None,
        on_close: Optional[Callable[[], None]] = None,
    ):
        self._url = url
        self._headers = {
            **headers,
            "Accept": "text/event-stream",
            "Cache-Control": "no-cache",
        }
        self._reconnect = reconnect
        self._reconnect_delay = reconnect_delay
        self._max_reconnect_delay = max_reconnect_delay
        self._on_open = on_open
        self._on_error = on_error
        self._on_close = on_close
        self._handlers: Dict[str, List[Callable]] = {}
        self._closed = False
        self._reconnect_attempts = 0
        self._thread: Optional[threading.Thread] = None

    def on(self, event_type: str, handler: Callable[[Dict[str, Any]], None]) -> None:
        self._handlers.setdefault(event_type, []).append(handler)

    def off(self, event_type: str, handler: Callable) -> None:
        if event_type in self._handlers:
            try:
                self._handlers[event_type].remove(handler)
            except ValueError:
                pass

    def close(self) -> None:
        self._closed = True
        if self._on_close:
            self._on_close()

    def start(self) -> "SSEStream":
        self._thread = threading.Thread(target=self._run, daemon=True)
        self._thread.start()
        return self

    def _emit(self, event_type: str, event: Dict[str, Any]) -> None:
        for handler in self._handlers.get(event_type, []):
            try:
                handler(event)
            except Exception:
                pass
        for handler in self._handlers.get("*", []):
            try:
                handler(event)
            except Exception:
                pass

    def _process_raw(self, raw: Dict[str, Any]) -> None:
        data_str = "\n".join(raw.get("data", []))
        if not data_str:
            return
        try:
            parsed = json.loads(data_str)
            data_field = parsed.get("data") or parsed
            event: Dict[str, Any] = {
                "id": raw.get("id") or parsed.get("id", ""),
                "type": parsed.get("type") or raw.get("type") or "unknown",
                "ts": parsed.get("ts", ""),
                "data": data_field,
                "smartSpaceId": parsed.get("smartSpaceId"),
                "runId": parsed.get("runId") or (data_field.get("runId") if isinstance(data_field, dict) else None),
                "entityId": parsed.get("entityId"),
                "agentEntityId": parsed.get("agentEntityId"),
                "seq": parsed.get("seq"),
            }
            self._emit(event["type"], event)
        except (json.JSONDecodeError, Exception):
            pass

    def _run(self) -> None:
        while not self._closed:
            try:
                with requests.get(
                    self._url,
                    headers=self._headers,
                    stream=True,
                    timeout=None,
                ) as resp:
                    if not resp.ok:
                        raise ConnectionError(
                            f"SSE connection failed: {resp.status_code} {resp.reason}"
                        )

                    self._reconnect_attempts = 0
                    if self._on_open:
                        self._on_open()

                    current: Dict[str, Any] = {"data": [], "id": None, "type": None}

                    for raw_line in resp.iter_lines(decode_unicode=True):
                        if self._closed:
                            return

                        if raw_line is None:
                            continue

                        line: str = raw_line

                        if line.startswith(":"):
                            continue

                        if line == "":
                            if current["data"]:
                                self._process_raw(current)
                            current = {"data": [], "id": None, "type": None}
                            continue

                        colon = line.find(":")
                        if colon == -1:
                            continue
                        field = line[:colon]
                        value = line[colon + 1 :].lstrip(" ")

                        if field == "id":
                            current["id"] = value
                        elif field == "event":
                            current["type"] = value
                        elif field == "data":
                            current["data"].append(value)

            except Exception as exc:
                if self._closed:
                    return
                if self._on_error:
                    self._on_error(exc)
                if not self._reconnect:
                    break
                self._reconnect_attempts += 1
                delay = min(
                    self._reconnect_delay * (2 ** (self._reconnect_attempts - 1)),
                    self._max_reconnect_delay,
                )
                time.sleep(delay)

File: react-native-sdk/src/sse.ts
import type { StreamEvent, StreamEventHandler, HsafaStream } from './types';

export interface SSEStreamOptions {
  url: string;
  headers: Record<string, string>;
  onEvent?: (event: StreamEvent) => void;
  onError?: (error: Error) => void;
  onOpen?: () => void;
  onClose?: () => void;
  reconnect?: boolean;
  reconnectDelay?: number;
  maxReconnectDelay?: number;
}

/**
 * SSE client for React Native using XMLHttpRequest.
 *
 * React Native does not support fetch streaming (getReader/ReadableStream),
 * so we use XHR with onprogress to process incoming chunks incrementally.
 * XMLHttpRequest is a React Native built-in — no extra dependencies needed.
 */
export class SSEStream implements HsafaStream {
  private xhr: XMLHttpRequest | null = null;
  private handlers: Map<string, Set<StreamEventHandler>> = new Map();
  private closed = false;
  private reconnectAttempts = 0;
  private lastIndex = 0;
  private buffer = '';
  private options: SSEStreamOptions;

  constructor(options: SSEStreamOptions) {
    this.options = options;
    this.connect();
  }

  on(event: string, handler: StreamEventHandler): void {
    if (!this.handlers.has(event)) {
      this.handlers.set(event, new Set());
    }
    this.handlers.get(event)!.add(handler);
  }

  off(event: string, handler: StreamEventHandler): void {
    this.handlers.get(event)?.delete(handler);
  }

  close(): void {
    this.closed = true;
    if (this.xhr) {
      this.xhr.abort();
      this.xhr = null;
    }
    this.options.onClose?.();
  }

  private emit(eventType: string, event: StreamEvent): void {
    this.handlers.get(eventType)?.forEach((handler) => handler(event));
    this.handlers.get('*')?.forEach((handler) => handler(event));
    this.handlers.get('hsafa')?.forEach((handler) => handler(event));
  }

  private connect(): void {
    if (this.closed) return;

    const xhr = new XMLHttpRequest();
    this.xhr = xhr;
    this.lastIndex = 0;
    this.buffer = '';

    xhr.open('GET', this.options.url, true);

    const allHeaders: Record<string, string> = {
      ...this.options.headers,
      'Accept': 'text/event-stream',
      'Cache-Control': 'no-cache',
    };
    for (const [key, value] of Object.entries(allHeaders)) {
      xhr.setRequestHeader(key, value);
    }

    xhr.onprogress = () => {
      if (this.closed) return;
      const newText = xhr.responseText.slice(this.lastIndex);
      this.lastIndex = xhr.responseText.length;
      this.processChunk(newText);
    };

    xhr.onload = () => {
      if (this.closed) return;
      // Process any remaining buffered data
      if (this.buffer.trim()) {
        this.processChunk('\n');
      }
      // Stream ended cleanly — reconnect
      if (this.options.reconnect !== false) {
        this.scheduleReconnect();
      }
    };

    xhr.onerror = () => {
      if (this.closed) return;
      const err = new Error('SSE connection failed');
      this.options.onError?.(err);
      this.handlers.get('error')?.forEach((handler) =>
        handler({ id: '', type: 'error', ts: new Date().toISOString(), data: { error: err.message } } as StreamEvent)
      );
      if (this.options.reconnect !== false) {
        this.scheduleReconnect();
      }
    };

    xhr.onreadystatechange = () => {
      if (xhr.readyState === XMLHttpRequest.OPENED) {
        this.reconnectAttempts = 0;
        this.options.onOpen?.();
      }
    };

    xhr.send();
  }

  private processChunk(chunk: string): void {
    this.buffer += chunk;
    const lines = this.buffer.split('\n');
    this.buffer = lines.pop() ?? '';

    let currentEvent: { id?: string; type?: string; data: string[] } = { data: [] };

    for (const line of lines) {
      if (line.startsWith(':')) {
        // Comment line (keepalive), ignore
        continue;
      }

      if (line === '') {
        // Empty line = end of SSE event block
        if (currentEvent.data.length > 0) {
          this.processEvent(currentEvent);
        }
        currentEvent = { data: [] };
        continue;
      }

      const colonIdx = line.indexOf(':');
      if (colonIdx === -1) continue;

      const field = line.slice(0, colonIdx);
      const value = line.slice(colonIdx + 1).trimStart();

      switch (field) {
        case 'id':
          currentEvent.id = value;
          break;
        case 'event':
          currentEvent.type = value;
          break;
        case 'data':
          currentEvent.data.push(value);
          break;
      }
    }
  }

  private processEvent(raw: { id?: string; type?: string; data: string[] }): void {
    const dataStr = raw.data.join('\n');
    if (!dataStr) return;

    try {
      const parsed = JSON.parse(dataStr);

      // Unwrap gateway envelope: { seq, smartSpaceId, entityId, ..., data: <actual event data> }
      const outer = parsed.data != null && typeof parsed.data === 'object' && !Array.isArray(parsed.data)
        ? parsed.data
        : {};
      const hasEnvelope = 'data' in outer && typeof outer.data === 'object' && outer.data !== null;
      const eventData = hasEnvelope ? outer.data : (parsed.data ?? parsed);

      const event: StreamEvent = {
        id: raw.id || parsed.id || '',
        type: parsed.type || raw.type || 'unknown',
        ts: parsed.ts || new Date().toISOString(),
        data: eventData,
        smartSpaceId: parsed.smartSpaceId || outer.smartSpaceId,
        runId: parsed.runId || outer.runId || eventData?.runId,
        entityId: parsed.entityId || outer.entityId || eventData?.entityId,
        entityType: parsed.entityType || outer.entityType,
        agentEntityId: parsed.agentEntityId || outer.agentEntityId || eventData?.agentEntityId,
        seq: parsed.seq || outer.seq,
      };

      this.options.onEvent?.(event);
      this.emit(event.type, event);
    } catch {
      // Ignore malformed events
    }
  }

  private scheduleReconnect(): void {
    if (this.closed) return;

    this.reconnectAttempts++;
    const baseDelay = this.options.reconnectDelay ?? 1000;
    const maxDelay = this.options.maxReconnectDelay ?? 30000;
    const delay = Math.min(baseDelay * Math.pow(2, this.reconnectAttempts - 1), maxDelay);

    setTimeout(() => {
      if (!this.closed) {
        this.connect();
      }
    }, delay);
  }
}

export function createSSEStream(options: SSEStreamOptions): SSEStream {
  return new SSEStream(options);
}
